vlib/x/json2/decoder2/decode.v

```
@@ -615,33 +705,149 @@ fn (mut decoder Decoder) decode_value[T](mut val T) ! {
+									current_field_info = current_field_info.next
+									continue
+								}
+							}
+							.number {
+								if decoder.json[decoder.current_node.next.value.position] == `0` {
+									if decoder.current_node.next.value.length == 1 {
+										current_field_info = current_field_info.next
+										continue
+									} else if decoder.current_node.next.value.length == 3 {
+										if unsafe {
+											vmemcmp(decoder.json.str +
+												decoder.current_node.next.value.position,
+												float_zero_in_string.str, float_zero_in_string.len) == 0
+										} {
+											current_field_info = current_field_info.next
+											continue
+										}
+									}
+								}
+							}
+							else {}
+						}
+					}
+
+					// check if the key matches the field name
+					if key_info.length - 2 == current_field_info.value.json_name_len {
 						if unsafe {
-							vmemcmp(decoder.json.str + key_info.position + 1, field.name.str,
-								field.name.len) == 0
+							vmemcmp(decoder.json.str + key_info.position + 1, current_field_info.value.json_name_ptr,
+								current_field_info.value.json_name_len) == 0
 						} {
-							$if field.typ is $option {
-								mut unwrapped_val := create_value_from_optional(val.$(field.name))
-								decoder.decode_value(mut unwrapped_val)!
-								val.$(field.name) = unwrapped_val
-							} $else {
-								decoder.decode_value(mut val.$(field.name))!
+							$for field in T.fields {
+								if field.name.len == current_field_info.value.field_name_len {
+									if unsafe {
+										(&u8(current_field_info.value.field_name_str)).vstring_with_len(field.name.len) == field.name
+									} {
+										// value node
+										decoder.current_node = decoder.current_node.next
+
+										if current_field_info.value.is_skip {
+											if current_field_info.value.is_required == false {
+												panic('This should not happen. `skip` field should not be processed here without be a `required` field')
```

```suggestion
												panic('This should not happen. `skip` field should not be processed here without a `required` field')
```

---

All the rest return errors, why does this panic?

---

Because this should never happen. However, if someone were to perform maintenance on this library one day, and come across this panic, they would know that something was done that shouldn't have been done.

---

You are already returning errors in other similar places.
```
	return error('`raw` attribute can only be used with string fields')
        return error('`raw` attribute cannot be used with enum fields')
        return error('`json` attribute must have an argument')
```
etc ... I think that is good, and you can return an error here as well. 
There is no reason for such a library to panic.

---

> However, if someone were to perform maintenance on this library one day
Leave that to the future. 
The maintainer can just as well search for the error message too.

---

You can also add a comment if you want, but imho it is not needed - the error message is clear.

---

Good. Done